skfp/model_selection/splitters/pubchem_split.py

```diff
@@ -0,0 +1,470 @@
+
+    response = None
+    trial = 0
+    while trial < n_retries:
+        try:
+            response = requests.get(url, timeout=20)
+            response_status = response.status_code
+            if response_status != 200 and verbosity > 0:
+                warnings.warn(
+                    f"PUG error for URL {url}, status code {response_status}, description: {response.text}"
+                )
+
+            if response:
+                response_json = response.json()
+                cid = response_json["IdentifierList"]["CID"][0]
+                cid = str(cid) if cid else None
+                return cid
+            else:
+                trial += 1
+
+        except (requests.exceptions.Timeout, requests.exceptions.RequestException):
+            time.sleep(1)
+            trial += 1
+        except (KeyError, ValueError):
+            return None
+    raise RuntimeError("CID cannot be downloaded, unknown PubChem error")
+
+
+def _get_earliest_publication_date(cid: Optional[str], n_retries: int) -> Optional[int]:
+    """
+    Get the date of the earliest publication from PubChem where a given molecule appears.
+    There are molecules without publications, for which we return None.
+    """
+    if not cid:
+        return None
+
+    query = {
+        "download": "*",
+        "collection": "literature",
+        "where": {"ands": [{"cid": str(cid)}]},
+        "order": ["articlepubdate,asc"],
+        "start": 0,
+        "limit": 3,  # fetch a few, just in case
+        "downloadfilename": f"pubchem_cid_{cid}_literature",
+        "nullatbottom": 1,
+    }
+    base_url = "https://pubchem.ncbi.nlm.nih.gov/sdq/sdqagent.cgi"
+    params = {"infmt": "json", "outfmt": "json", "query": json.dumps(query)}
+
+    def get_publication_date(response: list[dict]) -> Optional[int]:
+        article_pub_date = None
+        for entry in response:
+            if "articlepubdate" in entry:
+                return int(entry["articlepubdate"][:4])  # get year of publication
+        return article_pub_date
+
+    trial = 0
+    while trial < n_retries:
+        try:
+            response = requests.get(base_url, params=params, timeout=10)
+            status_code = response.status_code
+            response = response.json()
+
+            if status_code != 200:
+                f"PUG error for URL {base_url}, status code {status_code}, description: {response.text}"
+                trial += 1
+                continue
+
+            if isinstance(response, list) and len(response) >= 1:
+                date = get_publication_date(response)
+                return date
+            else:
+                # most probably "Server too busy" error
+                time.sleep(1)
+
+        except requests.ConnectionError:
+            # most probably unstable PubChem network
+            time.sleep(1)
+        except requests.exceptions.Timeout:
+            print(f"For cid: {cid}, request timeout was reached, trying again... ")
```

Comment:
Don't print, particularly since, this function is run in parallel, just quietly retry